filter.hs

```diff
@@ -0,0 +1,5 @@
+filter' :: (a -> Bool) -> [a] -> [a]
+filter' _ [] = []
+filter' p (x:xs)
+  | p x       = x : filter' p xs
+  | otherwise = filter' p xs
```

Comment:
```hs
*Main> filter (>3) [1,5,3,2,1,6,4,3,2,1]
[5,6,4]
*Main> filter' (>3) [1,5,3,2,1,6,4,3,2,1]
[5,6,4]
*Main> filter' (==3) [1,2,3,4,5]
[3]
*Main> filter even [1..10]
[2,4,6,8,10]
*Main> let notNull x = not (null x) in filter' notNull [[1,2,3],[],[3,4,5],[2,2],[],[],[]]
[[1,2,3],[3,4,5],[2,2]]
*Main> filter' (`elem` ['a'..'z']) "u LaUgH aT mE BeCaUsE I aM diFfeRent"
"uagameasadifeent"
*Main> filter' (`elem` ['A'..'Z']) "i LAuGh at you bEcause u R all the same"
"LAGER"
```